daemon/images/images.go

```diff
@@ -177,75 +177,75 @@ func (i *ImageService) Images(imageFilters filters.Args, all bool, withExtraAttr
 				if imageFilters.Contains("reference") { // skip images with no references if filtering by reference
 					continue
 				}
-				newImage.RepoDigests = []string{"<none>@<none>"}
-				newImage.RepoTags = []string{"<none>:<none>"}
+				summary.RepoDigests = []string{"<none>@<none>"}
+				summary.RepoTags = []string{"<none>:<none>"}
 			} else {
 				continue
 			}
-		} else if danglingOnly && len(newImage.RepoTags) > 0 {
+		} else if danglingOnly && len(summary.RepoTags) > 0 {
 			continue
 		}
 
 		if withExtraAttrs {
-			// lazily init variables
-			if imagesMap == nil {
+			// Lazily init summaryMap and allContainers
+			if summaryMap == nil {
 				allContainers = i.containers.List()
 				allLayers = i.layerStore.Map()
-				imagesMap = make(map[*image.Image]*types.ImageSummary)
+				summaryMap = make(map[*image.Image]*types.ImageSummary)
 				layerRefs = make(map[layer.ChainID]int)
 			}
 
 			// Get container count
-			newImage.Containers = 0
+			var containers int64
 			for _, c := range allContainers {
 				if c.ImageID == id {
-					newImage.Containers++
+					containers++
 				}
 			}
+			// NOTE: By default, Containers is -1, or "not set"
+			summary.Containers = containers
 
 			// count layer references
 			rootFS := *img.RootFS
 			rootFS.DiffIDs = nil
 			for _, id := range img.RootFS.DiffIDs {
 				rootFS.Append(id)
-				chid := rootFS.ChainID()
-				layerRefs[chid]++
-				if _, ok := allLayers[chid]; !ok {
-					return nil, fmt.Errorf("layer %v was not found (corruption?)", chid)
-				}
+				layerRefs[rootFS.ChainID()]++
 			}
-			imagesMap[img] = newImage
+			summaryMap[img] = summary
 		}
-
-		images = append(images, newImage)
+		summaries = append(summaries, summary)
 	}
 
 	if withExtraAttrs {
 		// Get Shared sizes
-		for img, newImage := range imagesMap {
+		for img, summary := range summaryMap {
 			rootFS := *img.RootFS
 			rootFS.DiffIDs = nil
 
-			newImage.SharedSize = 0
+			// Indicate that we collected shared size information (default is -1, or "not set")
+			summary.SharedSize = 0
 			for _, id := range img.RootFS.DiffIDs {
 				rootFS.Append(id)
```

Comment:
Unrelated to this patch, the `Append` was alread done when counting layer references ?

Reply:
What this does, is that it computes a chain ID of each layer.
The chain ID is equal to diff ID for the bottom-most layer and for the rest it is given by hashing the parent chain ID and the current diff ID.

The logic is quite obfuscated - I added this commit to clear it up https://github.com/moby/moby/pull/42578/commits/78e0c73e91ad5e57738531b26251e19802f9537a, but it will be merged in https://github.com/moby/moby/pull/42578 to keep the scope of this PR small(-er)

Reply:
yes, this logic is somewhat confusing; what it comes down to is that `rootFS.ChainID()` is not a simple "getter", but is _calculating_ the chain-id based on the DiffIDs; see https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/image/rootfs.go#L46-L53


This code is basically doing a "shallow" copy of `img.RootFS` (to copy all other properties _except for_ the DiffIDs), then adds the DiffIDs of the rootfs, and calls `rootFS.ChainID()` to calculate the ID.

Possibly we could call `layer.CreateChainID(img.RootFS.DiffIDs)` directly to get the chain-ID (after doing the windows-specific checks if needed), to make it clearer we're calculating something here.

(not for this PR though)